pgdog/src/backend/pool/dns_cache.rs

```diff
@@ -0,0 +1,388 @@
+use futures::future::join_all;
+use once_cell::sync::Lazy;
+use std::collections::{HashMap, HashSet};
+use std::net::IpAddr;
+use std::sync::{Arc, RwLock};
+use std::time::{Duration, Instant};
+use tokio::net::lookup_host;
+use tokio::select;
+use tokio::time::sleep;
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, info, warn};
+
+// -------------------------------------------------------------------------------------------------
+// ------ Constant(s) ------------------------------------------------------------------------------
+
+const DEFAULT_TTL: Duration = Duration::from_secs(30);
+
+// -------------------------------------------------------------------------------------------------
+// ------ Singleton(s) -----------------------------------------------------------------------------
+
+static DNS_CACHE: Lazy<Arc<DnsCache>> = Lazy::new(|| {
+    Arc::new(DnsCache {
+        cache: Arc::new(RwLock::new(HashMap::new())),
+        hostnames: Arc::new(RwLock::new(HashSet::new())),
+        ttl: Arc::new(RwLock::new(DEFAULT_TTL)),
+    })
+});
+
+// -------------------------------------------------------------------------------------------------
+// ------ DnsCache :: Public interface -------------------------------------------------------------
+
+#[derive(Clone, Debug)]
+struct DnsCacheEntry {
+    ip: IpAddr,
+    resolved_at: Instant,
+    ttl: Duration,
+}
+
+pub struct DnsCache {
+    cache: Arc<RwLock<HashMap<String, DnsCacheEntry>>>,
+    hostnames: Arc<RwLock<HashSet<String>>>,
+    ttl: Arc<RwLock<Duration>>,
+}
+
+impl DnsCache {
+    /// Retrieve the global DNS cache instance.
+    pub fn global() -> Arc<DnsCache> {
+        DNS_CACHE.clone()
+    }
+
+    /// Resolve hostname to socket address string.
+    pub async fn resolve(&self, hostname: &str) -> Result<IpAddr, std::io::Error> {
+        if let Some(ip) = self.get_cached_ip(hostname) {
+            return Ok(ip);
+        }
+
+        // Track hostname for future refreshes.
+        {
+            let mut hostnames = self.hostnames.write().unwrap();
+            if hostnames.insert(hostname.to_string()) {
+                info!("DNS Cache :: Added hostname: {}", hostname);
```

Comment:
This feels like a `debug!` level log. Maybe something like this to keep existing format:

```rust
debug!("dns cache added hostname \"{}\"",  hostname);
```